### PR TITLE
계산기 [STEP 2] MINT

### PR DESCRIPTION
**Pull request by mint3382**
@JoSH0318 
안녕하세요 밈! 목표했던 날보다 몹시도 늦은 step 2 pr입니다🥲
이번에도 잘 부탁드립니다~ 😎

# STEP 2
### 고차함수의 사용_ compactMap
프로젝트의 설명 중 고차함수를 최대한 사용해보라는 안내에 따라 `parse`함수 내에서 for 문 대신 `forEach`를 사용했습니다. 그러기 위해서 옵셔널 바인딩이 필요한 부분을 고민하게 되었는데 전에 옵셔널 스터디에서 했던 것이 생각나 `compactMap`으로 바인딩 해 주었습니다. 결론적으로 변수 선언하는 부분을 줄일 수 있었습니다.

### UML의 해석_ split, componentsByOperators
처음 주어진 UML을 온전하게 이해하는 것에 오랜 시간이 걸렸습니다. 특히 `split`과 `componentsByOperators` 함수를 이해하기가 쉽지 않았는데 `split`이 있는데 굳이 `componentsByOperators`가 있어야 하나는 생각을 했습니다. 온전히 저만의 생각은 아니지만 `componentsByOperators`를 구현함에 있어 `split`의 기준을 " " 공백으로 두었습니다. 이 때문에 위와 같은 생각이 더 커졌는데, 코드를 읽을때 조금 더 쉽게 읽기 위함으로 이해하고 넘어갔습니다.

### enum Test
`unit Test`를 하는 것에 있어서 `enum` 타입인 `ExpressionParser`나 `Operator`를 따로 `test` 할 수가 없는데 이 경우는 다른 `FormulaTests`에서 같이 해도 되는지 궁금합니다. `sut`로 부르지 않아도 사용할 수 있기에 `FormulaTests`의 `given` 형식으로만 사용해 놓았습니다.

### unit Test
이전 스텝에서는 `TDD`로 진행하였기에 `unit Test`를 최대한 자세하게 작성하였습니다. 이번 스텝의 경우는 `TDD`로 진행한 것이 아니라 단순히 `unit Test`의 기능을 사용하고 싶었던 건데 밈은 이 경우에도 최대한 자세하게 작성하시는지 궁금합니다. `test`를 어느 정도까지 진행해야 하는 것인지 아직 감이 잘 안 오는 것 같아요. 저는 이번 스텝에서는 구현을 한 후 `unit Test`로 생각을 검증만 하고 있기에 어느정도 간략하게 작성하였습니다.

### CODE REVIEW

File: Calculator/Calculator/Model/ExpressionParser.swift [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import Foundation
+
+enum ExpressionParser {
+    static func parse(from input: String) -> Formula {
+        var formula = Formula()
+        let components = componentsByOperators(from: input)
+        
+        components.filter { Double($0) != nil }
+            .compactMap { Double($0) }
+            .forEach { formula.operands.enqueue($0) }
```
Comment (JoSH0318) [inline Calculator/Calculator/Model/ExpressionParser.swift]: filter와 compactMap이 같은 역할을 하고 있는 것 같아요. 😀
때문에 filter는 불필요한 코드로 보여집니다. 🥹
> HINT: compactMap에서 Double 타입 변환에 실패하면 nil

File: Calculator/Calculator/Model/Fornula.swift [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        
+        while `operator` != nil {
+            guard let identifiedOperator = `operator` else { break }
+            guard let newOperand = operands.dequeue() else { break }
+            
+            result = identifiedOperator.calculate(lhs: result, rhs: newOperand)
+            `operator` = operators.dequeue()
+        }
+        
+        return result
```
Comment (JoSH0318) [inline Calculator/Calculator/Model/Fornula.swift]: 1. 저의 개인적인 견해로는 가독성이 떨어진다고 판단됩니다.
optional binding 과정을 포함하여서 변/상수가 너무 많아보여요. 🥹
코드를 줄여볼 수 있을 것 같아요.
> HINT: 만약 아래의 예시 코드와 같이 while 문을 구현한다면 어떻게 동작할까요?
```swift 
while let newOperator = operators.dequeue() {
    // ... some code
}
```

2. while 조건문에 `!= nil`을 사용하셨어요.  만약 지금과 같은 코드를 유지한다면 만들어 놓으신 `isEmpty`를 사용하는게 더 좋아보여요!

File: Calculator/Calculator/Model/Fornula.swift [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+import Foundation
+
+struct Formula {
+    var operators = CalculatorItemQueue<Operator>()
+    var operands = CalculatorItemQueue<Double>()
+    
+    mutating func result() -> Double {
+        var `operator` = operators.dequeue()
+        guard var result = operands.dequeue() else { return 0 }
```
Comment (JoSH0318) [inline Calculator/Calculator/Model/Fornula.swift]: 1. 연산 과정에서 operand로 0이 입력될 수도 있어요. 때문에 dequeue과정에서 0이 반환될 수도 있습니다.
만약 지금처럼 dequeue의 결과로 nil이라면, 0을 반환하게 되고 이것이 실제 dequeue된 값인지 optional binding의 return 결과인지 파악할 수가 없을 것 같아요! 😀
2. result로 선언했기 때문에 dequeue의 결과가 result가 되는 것 같이 느껴져요. 사실상 `result가 될 값, 중간값`이 맞을 것 같아요. 민트의 생각은 어떠한가요?

**Review by JoSH0318 — COMMENTED** [review]
안녕하세요👀 민트🍵

어려운 요구사항임에도 코드를 잘 작성해주신 것 같아요!!!
아이디어도 좋고, 컨벤션도 깔끔하고, 객체/함수 역할도 명확하네요!

몇 가지 코멘트, 요구사항 남겼습니다!

---

### 답변
1️⃣ enum Test
Q: unit Test를 하는 것에 있어서 enum 타입인 ExpressionParser나 Operator를 따로 test 할 수가 없는데 이 경우는 다른 FormulaTests에서 같이 해도 되는지 궁금합니다. sut로 부르지 않아도 사용할 수 있기에 FormulaTests의 given 형식으로만 사용해 놓았습니다.

> A: 요구사항에도 나와 있듯이 따로 해주시는 게 원칙적으로 맞다고 생각이 들어요! 충분히 따로 test를 해볼 수 있으니까요. 현재 ExpressionParser, Formula를 동시에 test하기 때문에 어떠한 것을 test하는지 명확하지 않은 부분이 있습니다. 따로 진행해주시면 좋을 것 같아요!

2️⃣ unit Test
Q: 이전 스텝에서는 TDD로 진행하였기에 unit Test를 최대한 자세하게 작성하였습니다. 이번 스텝의 경우는 TDD로 진행한 것이 아니라 단순히 unit Test의 기능을 사용하고 싶었던 건데 밈은 이 경우에도 최대한 자세하게 작성하시는지 궁금합니다. test를 어느 정도까지 진행해야 하는 것인지 아직 감이 잘 안 오는 것 같아요. 저는 이번 스텝에서는 구현을 한 후 unit Test로 생각을 검증만 하고 있기에 어느정도 간략하게 작성하였습니다.

> A: "Unit Test를 어느정도까지, 얼마나 디테일하게 작성해야하냐"가 질문의 요지인 것 같습니다. 맞나요🥲? 예를 들어 `3개씩은 test 해야한다`와 같은 정량적인 기준이 없기 때문에 명쾌한 답변이 어려울 것 같아요. 
다만 간결하되 최대한 모든 상황, 예외 상황을 test해보는 게 맞다고 생각이 들어요. 때문에 어떠한 상황이 있을 것인지 파악하고, 특히 edge case에는 무엇이 있을지 예측하고 이 모든 것을 test해야 좋다고 생각합니다.

---

### 요구사항 / 질문
1️⃣ Extension
현재 코드에는 Extension파일에 Double과 String Extension을 모아두셨어요. 만약 Extension한 타입이 많아진다면 어떨까요? 찾아보고 싶은, 수정하고 싶은 타입을 찾기 어려울 것 같아요. 
Extension(그룹) > String+(파일), Extension(그룹) > Double+(파일)과 같이 파일트리를 구성하면 더 좋을 것 같아요. 민트의 생각은 어떠신가요?

2️⃣ 예외 처리
요구사항에는 여러가지 예외처리를 할 것을 요구하고 있습니다. 
실행예시와 텍스트를 통해서 0으로 나눴을 경우, operand가 없는 경우(NaN)는 꼭 예외처리를 해줘야한다는 것을 알 수 있어요.
하지만 민트의 코드에 예외처리 부분이 조금 아쉬운 것 같아요. 🥹
unit test를 동반하여 예외처리를 구현해주셨으면 좋겠습니다.
> 우리가 실생활에서 사용하는 계산기의 경우  어떠한 방법으로 예외처리를 해주고 있는지 생각해보면 좋을 것 같아요.

File: Calculator/Calculator/Model/ExpressionParser.swift (hunk 1 above)
Comment (mint3382) [reply Calculator/Calculator/Model/ExpressionParser.swift]: filter 삭제하였습니다, 감사합니다!

File: Calculator/Calculator/Model/Fornula.swift (hunk 2 above)
Comment (mint3382) [reply Calculator/Calculator/Model/Fornula.swift]: while let 을 사용하여 바꾸며 `!= nil`은 삭제하였습니다~ 🤣

File: Calculator/Calculator/Model/Fornula.swift (hunk 3 above)
Comment (mint3382) [reply Calculator/Calculator/Model/Fornula.swift]: 1. error 타입을 생성하며 해당 문제 시 에러를 던지도록 하였습니다. 
2. while let 으로 바꾸며 네이밍도 수정되었습니다!

**Comment by mint3382** [comment]
### Error
에러 타입을 만들어서 0으로 나눌 때나 수식이 완성되지 못했을 경우 에러를 던지도록 코드를 수정하였습니다~!

### 마지막에 빈 String이 들어가는 경우
test에 edge case들을 추가하다가 마지막에 빈 String이 들어가는 경우 오류가 발생하는 것을 알게되어 parse에서 filter를 통해 거를 때 빈 String도 거를 수 있게 변경하였습니다. 


😈
